KAFKA-5188: Integration tests for transactions

apurvam opened the pull request.

hachikuji committed:
Add transactions integration tests

apurvam commented:
cc @hachikuji

hachikuji commented on core/src/test/scala/integration/kafka/api/TransactionsTest.scala:
```diff
@@ -0,0 +1,314 @@
+
+      consumer.subscribe(List(topic1, topic2))
+
+      val records = pollUntilExactlyNumRecords(consumer, 2)
```
How can we be sure that the aborted records wouldn't be returned if we waited a little longer? Maybe we should do the aborted transaction first to ensure that it. Also, it might make sense to have a second READ_UNCOMMITTED consumer and verify that all records are returned.

apurvam replied:
I did both: moved the abort to be before the commit, and also added a `READ_UNCOMMITTED` consumer to ensure that it reads all 4 records.

hachikuji commented on core/src/test/scala/integration/kafka/api/TransactionsTest.scala:
```diff
@@ -0,0 +1,314 @@
+      producer2.send(producerRecord(topic2, "2", "4", willBeCommitted = true))
+
+      try {
+        producer1.commitTransaction()
```
Not sure how exhaustive we need to be, but it might also be useful to also check fencing for the case that the zombie is in between transactions (i.e. the fenced operation should be the `send`).

apurvam replied:
I added test cases to check for fencing on send and also test fencing on add partitions.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/TransactionManager.java:
```diff
@@ -261,6 +261,23 @@ public synchronized boolean maybeSetError(Exception exception) {
+        if (isInErrorState() && requestHandler.isEndTxn()) {
+            // We shouldn't terminate abort requests from error states.
+            EndTxnHandler endTxnHandler = (EndTxnHandler) requestHandler;
+            if (endTxnHandler.builder.result() == TransactionResult.ABORT)
```
Seems like this was a bug fix. Did we have any unit tests which hit this path?

apurvam replied:
I just added one.

hachikuji commented on core/src/test/scala/integration/kafka/api/TransactionsTest.scala:
```diff
@@ -0,0 +1,314 @@
+        producer.beginTransaction()
+        val shouldCommit = if (random.nextInt() % 2 == 0) true else false
+
+        val records = pollUntilAtLeastNumRecords(consumer, Math.min(10, 1000 - recordsProcessed))
```
Seems like there's a good chance that all 1000 records will be returned on the first poll. Perhaps we can just divide the commit and abort cases into two separate tests so that we're sure we hit them both?

apurvam replied:
The reason to mix the commits and aborts in this test is to test the fact that offsets are actually committed or aborted with the the transaction. If offsets are not committed with a transaction commit, then we will get duplicates. If offsets are not aborted with an abort, we will lose data. That's why we need both. 

Is there a way to tune the consumer to receive a small number of messages at a time? 

Also, empirically on my machine, it looks like we fetch around 50 at a time, which is a good amount.

apurvam replied:
I explicitly set the `maxPollRecords` consumer config to be 50 for this test.

Commit:
WIP: added an uncomitted consumer to ensure that it reads all messages

Commit:
Addressed PR Comments:

1. Added test cases for fencing at different stages (test fail right now
because of bugs).
2. Added a test case for aborting from error state.
3. Made the sendOffsets test more robust.

Commit:
Undo changes to transactioncoordinator

Commit:
Rebased onto trunk, and fixed what a bug in the existing TC

apurvam commented on core/src/main/scala/kafka/coordinator/transaction/TransactionMetadata.scala:
```diff
@@ -135,7 +135,7 @@ private[transaction] class TransactionMetadata(val producerId: Long,
 
   def removePartition(topicPartition: TopicPartition): Unit = {
-    if (pendingState.isDefined || (state != PrepareCommit && state != PrepareAbort))
+    if (pendingState.isDefined && (state != PrepareCommit && state != PrepareAbort))
```
@guozhangwang does this look right to you? The basic integration test would hit this exception from the `WriteTxnMarkerRequestCompletionHandler` otherwise.

guozhangwang replied:
Hmm, I think it is not expected if `state != PrepareCommit && state != PrepareAbort` itself, what state it was when hitting this exception?

apurvam replied:
This is the stack trace when we hit the exception:

```[2017-05-12 18:15:55,481] ERROR Uncaught error in request completion: (org.apache.kafka.clients.NetworkClient:379)
java.lang.IllegalStateException: Transation metadata's current state is PrepareAbort, and its pending state is PrepareAbort while trying to remove partitions whose txn marker has been sent, this is not expected
    at kafka.coordinator.transaction.TransactionMetadata.removePartition(TransactionMetadata.scala:139)
    at kafka.coordinator.transaction.TransactionMarkerRequestCompletionHandler$$anonfun$onComplete$6$$anonfun$apply$4.apply(TransactionMarkerRequestCompletionHandler.scala:87)
    at kafka.coordinator.transaction.TransactionMarkerRequestCompletionHandler$$anonfun$onComplete$6$$anonfun$apply$4.apply(TransactionMarkerRequestCompletionHandler.scala:83)
    at scala.collection.TraversableLike$WithFilter$$anonfun$foreach$1.apply(TraversableLike.scala:772)
    at scala.collection.Iterator$class.foreach(Iterator.scala:727)
    at scala.collection.AbstractIterator.foreach(Iterator.scala:1157)
    at scala.collection.IterableLike$class.foreach(IterableLike.scala:72)
    at scala.collection.AbstractIterable.foreach(Iterable.scala:54)
    at scala.collection.TraversableLike$WithFilter.foreach(TraversableLike.scala:771)
    at 
````

apurvam replied:
It looks like it is hit after the transaction markers are written.

guozhangwang replied:
It is indeed a bug, I have fixed it in another PR.

apurvam commented on core/src/test/scala/integration/kafka/api/TransactionsTest.scala:
```diff
@@ -0,0 +1,451 @@
+    }
+  }
+
+  @Ignore @Test
```
This should be fixed by https://issues.apache.org/jira/browse/KAFKA-5231 

cc @guozhangwang

Commit:
Fix the send offsets test. We need to recreate a new consumer instance after every abort so that it gets the correct position from the server

apurvam commented on core/src/test/scala/integration/kafka/api/TransactionsTest.scala:
```diff
@@ -0,0 +1,450 @@
+    }
+  }
+
+  @Ignore @Test
```
This should be fixed with https://github.com/apache/kafka/pull/2979

Commit:
Rename calls to transactionManagre.hasPid to transactionManager.hasProducerId in new tests to reflect upstream changes

Commit:
Implement the bounce test

Commit:
Minor changes

Commit:
Rebase onto trunk and un-ignore one of the tests which now passes

Commit:
Got the bounce test to work

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/Sender.java:
```diff
@@ -322,23 +333,26 @@ private boolean maybeSendTransactionalRequest(long now) {
                 }
                 if (targetNode != null) {
                     if (nextRequestHandler.isRetry()) {
+                        log.trace("waiting " + retryBackoffMs + "ms before resending a transactional request: " + nextRequestHandler.requestBuilder());
```
nit: seems we usually capitalize the first word in log messages. Would be nice to be consistent. Also, we should use the `{}` substitutions.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/Sender.java:
```diff
@@ -322,23 +333,26 @@ private boolean maybeSendTransactionalRequest(long now) {
 
-        if (targetNode == null)
-            transactionManager.retry(nextRequestHandler);
+        transactionManager.retry(nextRequestHandler);
```
Was this a bug?

apurvam replied:
Yes. The `networkClientUtils.awaitReady` and `awaitLeastLoadedNodeReady` throws an IOException if the connection to the node fails. So we would have a `targetNode` which is the coordinator, and then when we tried to connect to it, it would throw an IOException. 

So `targetNode` would not be null, the transactional request would not be sent, and it wouldn't be reenqueued either.. so the top level methods like `commitTransaction`, `sendOffsets` etc., would wait forever.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/Sender.java:
```diff
@@ -281,27 +284,35 @@ private long sendProducerData(long now) {
+            if (!accumulator.flushInProgress())
+                accumulator.beginFlush();
+            transactionManager.reenqueue(nextRequestHandler);
+            log.trace("Going to wait for pending ProducerBatches to flush before sending an end transaction request");
```
nit: we should get in the habit of including the transactionalId in the messages which relate to transactions. That will make debugging easier when we have multiple producers in the same application.

hachikuji commented on core/src/main/scala/kafka/coordinator/transaction/TransactionCoordinator.scala:
```diff
@@ -390,6 +390,7 @@ class TransactionCoordinator(brokerId: Int,
                   // the log append was successful
                   responseCallback(Errors.NONE)
 
+
```
nit: seems unneeded?

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/TransactionManager.java:
```diff
@@ -246,7 +247,7 @@ public boolean isInTransaction() {
 
     public boolean isInErrorState() {
-        return currentState == State.ERROR;
+        return currentState == State.ERROR || currentState == State.FENCED;
```
Yikes. Do we have a unit test for this?

apurvam replied:
I added an assertion for this to an existing test.

hachikuji commented on core/src/test/scala/integration/kafka/api/TransactionsTest.scala:
```diff
@@ -0,0 +1,397 @@
+        assertTrue(expectedValues.contains(recordValue(record)))
+      }
+    } catch {
+      case e @ (_ : KafkaException | _ : ProducerFencedException) =>
```
nit: do we need this? An unexpected exception will cause the test to fail anyway. Same for the other tests.

hachikuji reviewed (commented):
Test cases are looking good! Left a few minor comments.

hachikuji commented on core/src/test/scala/integration/kafka/api/TransactionsBounceTest.scala:
```diff
@@ -0,0 +1,207 @@
+    recordValue.replace("-committed", "").replace("-aborted", "").toInt
+  }
+
+  private def producerRecord(topic: String, consumerRecord: ConsumerRecord[Array[Byte], Array[Byte]], willBeCommitted: Boolean) = {
```
Maybe instead of all the string munging, we could use the message key (or dare I suggest, a header?) to indicate whether the record should be committed or aborted.

apurvam replied:
ack. I updated the patch to use headers.

Commit:
Address PR Comments

Commit:
Use record headers to indicate expected committed or aborted status of transactional messages

hachikuji commented on core/src/test/scala/integration/kafka/api/TransactionsTest.scala:
```diff
@@ -0,0 +1,356 @@
+  }
+
+  @Ignore @Test
+  def testFencingOnSend() {
```
Another case is fencing on `sendOffsets`. I'll leave it up to you whether it's worthwhile or not.

apurvam replied:
I added this case as well.

Commit:
Test fencing on sendOffsets

hachikuji reviewed (approved):
LGTM

Commit:
Move try/catch out of the while loop in TransactionsBounceTest. Add logging to transaction coordinator when we hit an invalid state

Commit:
Ignore the bounce test

hachikuji commented:
Thanks for the updates. I will merge this as soon as the builds complete.
